gadget/update.go

```diff
@@ -342,44 +382,36 @@ func KernelUpdatePolicy(from, to *LaidOutStructure) (bool, ResolvedContentFilter
 	return false, nil
 }
 
-func resolveUpdate(oldVol *PartiallyLaidOutVolume, newVol *LaidOutVolume, policy UpdatePolicyFunc) (updates []updatePair, err error) {
+func resolveUpdate(oldVol *PartiallyLaidOutVolume, newVol *LaidOutVolume, policy UpdatePolicyFunc, newGadgetRootDir, newKernelRootDir string, kernelInfo *kernel.Info) (updates []updatePair, err error) {
 	if len(oldVol.LaidOutStructure) != len(newVol.LaidOutStructure) {
 		return nil, errors.New("internal error: the number of structures in new and old volume definitions is different")
 	}
 	for j, oldStruct := range oldVol.LaidOutStructure {
 		newStruct := newVol.LaidOutStructure[j]
-		// update only when new edition is higher than the old one; boot
-		// assets are assumed to be backwards compatible, once deployed
-		// are not rolled back or replaced unless a higher edition is
-		// available
+		// update only when the policy says so; boot assets
+		// are assumed to be backwards compatible, once
+		// deployed are not rolled back or replaced unless
+		// told by the new policy
 		if update, filter := policy(&oldStruct, &newStruct); update {
+
+			// ensure content is resolved and filtered at
+			// this point
+			resolvedContent, err := resolveVolumeContent(newGadgetRootDir, newKernelRootDir, kernelInfo, &newStruct, filter)
+			if err != nil {
+				return nil, err
+			}
+			newVol.LaidOutStructure[j].ResolvedContent = resolvedContent
+
```

Comment:
if filter != nil and len(resolvedContent) == 0  we should continue here I think (to match the original behavior when filter was considering the actual KernelUpdate flag)

Reply:
Thanks for this suggestion, I added a check for this now that continues if no resolvedContent or LaidOutContent) is present anymore.